src/storage/gcs/storage.c

```diff
@@ -0,0 +1,949 @@
+            cryptoError(
+                EVP_DigestSignUpdate((EVP_MD_CTX *)sign, (unsigned char *)strZ(result), (unsigned int)strSize(result)) <= 0,
+                "unable to update");
+
+            size_t signatureLen = 0;
+            cryptoError(EVP_DigestSignFinal((EVP_MD_CTX *)sign, NULL, &signatureLen) <= 0, "unable to get size");
+
+            Buffer *signature = bufNew(signatureLen);
+            bufUsedSet(signature, bufSize(signature));
+
+            cryptoError(EVP_DigestSignFinal((EVP_MD_CTX *)sign, bufPtr(signature), &signatureLen) <= 0, "unable to finalize");
+
+            // Add signature
+            strCatChr(result, '.');
+            strCatEncode(result, encodeBase64Url, signature);
+        }
+        FINALLY()
+        {
+            BIO_free((BIO *)bio);
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+            EVP_MD_CTX_cleanup((EVP_MD_CTX *)sign);
+#else
+            EVP_MD_CTX_free((EVP_MD_CTX *)sign);
+#endif
+            EVP_PKEY_free((EVP_PKEY *)privateKey);
+        }
+        TRY_END();
+    }
+    MEM_CONTEXT_TEMP_END();
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+typedef struct
+{
+    String *tokenType;
+    String *token;
+    time_t timeExpire;
+} StorageGcsAuthTokenResult;
+
+static StorageGcsAuthTokenResult
+storageGcsAuthToken(StorageGcs *this, time_t timeBegin)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(STORAGE_GCS, this);
+        FUNCTION_TEST_PARAM(TIME, timeBegin);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+    ASSERT(timeBegin > 0);
+
+    StorageGcsAuthTokenResult result = {0};
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        String *content = strNewFmt(
+            "grant_type=urn%%3Aietf%%3Aparams%%3Aoauth%%3Agrant-type%%3Ajwt-bearer&assertion=%s",
+            strZ(storageGcsAuthJwt(this, timeBegin)));
+
+        HttpHeader *header = httpHeaderNew(NULL);
+        httpHeaderAdd(header, HTTP_HEADER_HOST_STR, httpUrlHost(this->authUrl));
+        httpHeaderAdd(header, HTTP_HEADER_CONTENT_TYPE_STR, HTTP_HEADER_CONTENT_TYPE_APP_FORM_URL_STR);
+        httpHeaderAdd(header, HTTP_HEADER_CONTENT_LENGTH_STR, strNewFmt("%zu", strSize(content)));
+
+        HttpRequest *request = httpRequestNewP(
+            this->authClient, HTTP_VERB_POST_STR, httpUrlPath(this->authUrl), NULL, .header = header, .content = BUFSTR(content));
+        HttpResponse *response = httpRequestResponse(request, true);
+
+        KeyValue *kvResponse = jsonToKv(strNewBuf(httpResponseContent(response)));
+
+        // Check for an error
+        const String *error = varStr(kvGet(kvResponse, GCS_JSON_ERROR_VAR));
+
+        if (error != NULL)
+        {
+            const String *description = varStr(kvGet(kvResponse, GCS_JSON_ERROR_DESCRIPTION_VAR));
+            CHECK(description != NULL);
+
+            THROW_FMT(ProtocolError, "unable to get authentication token: [%s] %s", strZ(error), strZNull(description));
```

Comment:
Line 233 uses `strZNull(description)` but in line 231, there is a CHECK assertion that will error if the description is NULL - why do we need the CHECK? Using strZNull means we don't care if it is NULL, and we're going to immediately throw an error anyway, so I fear if we throw the CHECK error, we'll miss what was in the contents of `error` string.

Reply:
Yeah, I was originally going to allow it to be NULL and then changed my mind. Fixed in 1dc661ad7aeac8.

Reply:
That wasn't quite my point - I was suggesting removing the CHECK completely and just leaving the strZNull in the THROW `THROW_FMT(ProtocolError, "unable to get authentication token: [%s] %s", strZ(error), strZNull(description));`

Reply:
Ah, yeah, caffeine levels were clearly deficient.  Done as you suggest in 5f5f8c7422f9afc0228892.